### PR TITLE
KAFKA-14586: Moving StreamResetter to tools

### PR DESCRIPTION
**Pull request by vamossagar12**
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by vamossagar12** [comment]
@mjsax , I noticed you had worked on the StreamsResetter tool. Would you be able to review this smallish PR? The idea is to move it from core module to tools module which is part of this: https://issues.apache.org/jira/browse/KAFKA-14525

**Comment by vamossagar12** [comment]
Adding @fvaleri . There are checkstyle issues, which I would fix.

File: build.gradle
Comment (cadonna) [inline build.gradle]: Ticket https://issues.apache.org/jira/browse/KAFKA-14525 (the parent of https://issues.apache.org/jira/browse/KAFKA-14586) says 

> tools that don't require access to `core` classes and communicate via the kafka protocol (typically by using the client classes) should be moved to the `tools` module.

This addition contradicts that requirement.

As far as I see, `kafka.utils.CommandLineUtils` is the only dependency to `core`. Is that true? Would it be possible to move `kafka.utils.CommandLineUtils` to a different module? Or should we even get completely rid of that dependency by rewriting it in java and put it in a different module.

https://issues.apache.org/jira/browse/KAFKA-14576 should have a similar issue since also the console consumer should be moved to tools and it has a dependency to  `kafka.utils.CommandLineUtils`.

**Review by cadonna — COMMENTED** [review]
@vamossagar12 Thanks for the PR!

I had one major comment.

File: build.gradle
Comment (clolov) [reply build.gradle]: Hello, hello, chipping in to say this is the same conversation we had with @fvaleri in https://github.com/apache/kafka/pull/13122. There should be a separate PR soon containing changes for CommandLineUtils (et al.) to unblock all JIRA tickets depending on it 😊

Comment (vamossagar12) [reply build.gradle]: Thanks @clolov i was about to type the same thing :D So this PR and I guess your PR is blocked until https://github.com/apache/kafka/pull/13122 is merged?

Comment (vamossagar12) [reply build.gradle]: Also, not sure if this comment from Ismael has any relevance here: https://github.com/apache/kafka/pull/13095#issuecomment-1376021193?

Comment (cadonna) [reply build.gradle]: Ah, great! Thanks for sharing! Good to know that people is aware!

> Also, not sure if this comment from Ismael has any relevance here: https://github.com/apache/kafka/pull/13095#issuecomment-1376021193?

I think there they talk about the dependencies of tools' test module not of the tools per se.

Comment (vamossagar12) [reply build.gradle]: >I think there they talk about the dependencies of tools' test module not of the tools per se.

Oh Ok. Got it. Thanks.

Comment (fvaleri) [reply build.gradle]: Yes I'm working on https://issues.apache.org/jira/browse/KAFKA-14628 and really close to opening the PR.

Comment (fvaleri) [reply build.gradle]: @cadonna yes, there is some confusion on which tools need to be migrated. For example, `DumpLogSegments` has multiple dependencies on `core` classes, so it should be excluded, but was added as a sub task. Checking with @mimaison how we want to precede with these. All the rest can be migrated.

Comment (vamossagar12) [reply build.gradle]: btw, @fvaleri , unrelated to this PR, but there seems to another core dependency from `ReplicaVerificationTools`. Plz check my comment on https://issues.apache.org/jira/browse/KAFKA-14583. I can file an issue and migrate that one as well if you can also confirm the same. Thanks

Comment (vamossagar12) [reply build.gradle]: hi @cadonna , The PR for moving `CommandLineUtils` has been [merged to trunk](https://github.com/apache/kafka/pull/13131), meaning I don't need to include this dependency anymore. I have updated the same.

**Comment by vamossagar12** [comment]
> @vamossagar12 Thanks for the PR!
> 
> I had one major comment.

Thanks @cadonna . I think that comment has been addressed now based on the conversation.

**Comment by vamossagar12** [comment]
@cadonna , bumping this thread again.

**Comment by fvaleri** [comment]
@vamossagar12 as you can see from the CI output, there are checkstyle errors. Have you tried to run at least one of the system tests that are using the tool?

**Comment by vamossagar12** [comment]
> @vamossagar12 as you can see from the CI output, there are checkstyle errors. Have you tried to run at least one of the system tests that are using the tool?

hi @fvaleri , yes I am aware of the checkstyle issue and pointed it out here: https://github.com/apache/kafka/pull/13127#issuecomment-1396439952. I have run the StreamResetter tests and also modified the system test.

**Comment by cadonna** [comment]
@vamossagar12 The PR LGTM except for the checkstyle issues. Could you also please post a link to a successful system test run of the system tests that use the streams resetter?

**Comment by vamossagar12** [comment]
Thanks @cadonna , would do the same.

**Comment by vamossagar12** [comment]
@cadonna , @fvaleri I ran the system test for StreamsResetter from my local (and fixed the checkstyle). The tests are running, would keep an eye on it =>

```
./gradlew clean systemTestLibs                                                                                         

> Configure project :
Starting build with version 3.5.0-SNAPSHOT (commit id af3bedfd) using Gradle 7.6, Java 1.8 and Scala 2.13.10
Build properties: maxParallelForks=8, maxScalacThreads=8, maxTestRetries=0

BUILD SUCCESSFUL in 5m 44s
220 actionable tasks: 212 executed, 8 up-to-date
(ducktape) sarao@C02GG1KCML7H kafka % TC_PATHS="tests/kafkatest/tests/streams/streams_optimized_test.py::StreamsOptimizedTest" bash tests/docker/run_tests.sh

> Configure project :
Starting build with version 3.5.0-SNAPSHOT (commit id af3bedfd) using Gradle 7.6, Java 1.8 and Scala 2.13.10
Build properties: maxParallelForks=8, maxScalacThreads=8, maxTestRetries=0

BUILD SUCCESSFUL in 9s
172 actionable tasks: 172 up-to-date
docker exec ducker01 bash -c "cd /opt/kafka-dev && ducktape --cluster-file /opt/kafka-dev/tests/docker/build/cluster.json  ./tests/kafkatest/tests/streams/streams_optimized_test.py::StreamsOptimizedTest "
/usr/local/lib/python3.9/dist-packages/paramiko/transport.py:236: CryptographyDeprecationWarning: Blowfish has been deprecated
  "class": algorithms.Blowfish,
[INFO:2023-02-10 01:05:18,393]: starting test run with session id 2023-02-10--002...
[INFO:2023-02-10 01:05:18,395]: running 1 tests...
[INFO:2023-02-10 01:05:18,398]: Triggering test 1 of 1...
[INFO:2023-02-10 01:05:18,454]: RunnerClient: Loading test {'directory': '/opt/kafka-dev/tests/kafkatest/tests/streams', 'file_name': 'streams_optimized_test.py', 'cls_name': 'StreamsOptimizedTest', 'method_name': 'test_upgrade_optimized_topology', 'injected_args': {'metadata_quorum': 'REMOTE_KRAFT'}}
[INFO:2023-02-10 01:05:18,480]: RunnerClient: kafkatest.tests.streams.streams_optimized_test.StreamsOptimizedTest.test_upgrade_optimized_topology.metadata_quorum=REMOTE_KRAFT: on run 1/1
[INFO:2023-02-10 01:05:18,498]: RunnerClient: kafkatest.tests.streams.streams_optimized_test.StreamsOptimizedTest.test_upgrade_optimized_topology.metadata_quorum=REMOTE_KRAFT: Setting up...
[INFO:2023-02-10 01:05:18,503]: RunnerClient: kafkatest.tests.streams.streams_optimized_test.StreamsOptimizedTest.test_upgrade_optimized_topology.metadata_quorum=REMOTE_KRAFT: Running...
[INFO:2023-02-10 01:08:00,049]: RunnerClient: kafkatest.tests.streams.streams_optimized_test.StreamsOptimizedTest.test_upgrade_optimized_topology.metadata_quorum=REMOTE_KRAFT: Tearing down...
[INFO:2023-02-10 01:08:26,433]: RunnerClient: kafkatest.tests.streams.streams_optimized_test.StreamsOptimizedTest.test_upgrade_optimized_topology.metadata_quorum=REMOTE_KRAFT: PASS
[INFO:2023-02-10 01:08:26,437]: RunnerClient: kafkatest.tests.streams.streams_optimized_test.StreamsOptimizedTest.test_upgrade_optimized_topology.metadata_quorum=REMOTE_KRAFT: Data: None
================================================================================
SESSION REPORT (ALL TESTS)
ducktape version: 0.11.3
session_id:       2023-02-10--002
run time:         3 minutes 8.101 seconds
tests run:        1
passed:           1
flaky:            0
failed:           0
ignored:          0
================================================================================
test_id:    kafkatest.tests.streams.streams_optimized_test.StreamsOptimizedTest.test_upgrade_optimized_topology.metadata_quorum=REMOTE_KRAFT
status:     PASS
run time:   3 minutes 7.954 seconds
--------------------------------------------------------------------------------
```

**Comment by vamossagar12** [comment]
Tests seemed to have passed for one of the builds. The tests that failed (even the one is Streams) seem unrelated to me.

File: build.gradle
Comment (fvaleri) [inline build.gradle]: This is duplicated and the one below this seems to be unnecessary.

File: checkstyle/import-control.xml
Comment (fvaleri) [inline checkstyle/import-control.xml:489]: This subpackage is not needed if you move StreamsResetterTest to tools test.

File: streams/src/test/java/org/apache/kafka/streams/tools/StreamsResetterTest.java
Comment (fvaleri) [inline streams/src/test/java/org/apache/kafka/streams/tools/StreamsResetterTest.java:19]: I think this test belongs to the tools module, where the class under test lives.

**Review by fvaleri — CHANGES_REQUESTED** [review]
Hi @vamossagar12, thanks for fixing the checkstyle errors. 

I left a few comments. I also think that it would be better to use the shared CommandDefaultOptions class, like in all the other joptsimple tools (help and version code is duplicated here).

The system test looks good:

```sh
test_id:    kafkatest.tests.streams.streams_optimized_test.StreamsOptimizedTest.test_upgrade_optimized_topology.metadata_quorum=REMOTE_KRAFT
status:     PASS
run time:   1 minute 12.791 seconds
```

File: streams/src/test/java/org/apache/kafka/streams/tools/StreamsResetterTest.java
Comment (vamossagar12) [reply streams/src/test/java/org/apache/kafka/streams/tools/StreamsResetterTest.java:19]: Yeah even I had thought so. But the original layout was also similar (i.e StreamsResetter was under tools and the test lived here) so I didn't change it. @cadonna , WDYT?

File: build.gradle
Comment (vamossagar12) [reply build.gradle]: done, removed. Thanks for pointing this out.

File: checkstyle/import-control.xml
Comment (vamossagar12) [reply checkstyle/import-control.xml:489]: I guess this is related to my comment https://github.com/apache/kafka/pull/13127#discussion_r1104203628?

**Comment by vamossagar12** [comment]
Thanks for the review @fvaleri . I addressed/responded to the comments. Regarding moving to `CommandDefaultOptions`, is it better if we do it in a follow up PR as that is not directly connected to migrating to tools module. I am not saying it isn't important but trying to keep the scope of this PR limited to the migration since  StreamsResetter has been working this way since the beginning. WDYT?

File: streams/src/test/java/org/apache/kafka/streams/tools/StreamsResetterTest.java
Comment (cadonna) [reply streams/src/test/java/org/apache/kafka/streams/tools/StreamsResetterTest.java:19]: I would also move it to the tools module. 
@mjsax @guozhangwang Was there a specific reason not to have the test in the same module as the `StreamsResetter`?

Comment (vamossagar12) [reply streams/src/test/java/org/apache/kafka/streams/tools/StreamsResetterTest.java:19]: Got it @cadonna . Would wait for the response. Even I think it can move but let's see what guozhang and Mathias have to say.
Also, while we are at it, would like to know your opinion on https://github.com/apache/kafka/pull/13127#issuecomment-1427653494

**Comment by fvaleri** [comment]
> Regarding moving to CommandDefaultOptions, is it better if we do it in a follow up PR as that is not directly connected to migrating to tools module.

Hi @vamossagar12, I'm ok with doing it in a follow up PR. Do we have a Jira for that?

**Comment by vamossagar12** [comment]
> > Regarding moving to CommandDefaultOptions, is it better if we do it in a follow up PR as that is not directly connected to migrating to tools module.
> 
> Hi @vamossagar12, I'm ok with doing it in a follow up PR. Do we have a Jira for that?

Thanks @fvaleri , here you go: https://issues.apache.org/jira/browse/KAFKA-14734

**Review by fvaleri — APPROVED** [review]
LGTM. Thanks!

**Comment by vamossagar12** [comment]
Thanks @fvaleri !

**Comment by vamossagar12** [comment]
@cadonna , bumping this one again :)

File: streams/src/test/java/org/apache/kafka/streams/tools/StreamsResetterTest.java
Comment (cadonna) [reply streams/src/test/java/org/apache/kafka/streams/tools/StreamsResetterTest.java:19]: I would not wait for the answer. We can do that in a follow-up PR. Could you open a ticket for the move?

**Review by cadonna — APPROVED** [review]
Thanks @vamossagar12 !

LGTM!

**Comment by cadonna** [comment]
I restarted the build. Once the results are acceptable, I will merge the PR.

File: streams/src/test/java/org/apache/kafka/streams/tools/StreamsResetterTest.java
Comment (vamossagar12) [reply streams/src/test/java/org/apache/kafka/streams/tools/StreamsResetterTest.java:19]: Thanks @cadonna , here's the follow up ticket: https://issues.apache.org/jira/browse/KAFKA-14851

**Comment by cadonna** [comment]
Test failures are unrelated:
```
    Build / JDK 17 and Scala 2.13 / kafka.zk.ZkMigrationIntegrationTest.[1] Type=ZK, Name=testDualWrite, MetadataVersion=3.4-IV0, Security=PLAINTEXT
    Build / JDK 17 and Scala 2.13 / kafka.zk.ZkMigrationIntegrationTest.[1] Type=ZK, Name=testDualWrite, MetadataVersion=3.4-IV0, Security=PLAINTEXT
    Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.DedicatedMirrorIntegrationTest.testMultiNodeCluster()
    Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.DedicatedMirrorIntegrationTest.testMultiNodeCluster()
    Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.integration.ExactlyOnceSourceIntegrationTest.testConnectorBoundary
    Build / JDK 11 and Scala 2.13 / kafka.zk.ZkMigrationIntegrationTest.[1] Type=ZK, Name=testDualWrite, MetadataVersion=3.4-IV0, Security=PLAINTEXT
    Build / JDK 11 and Scala 2.13 / kafka.zk.ZkMigrationIntegrationTest.[1] Type=ZK, Name=testDualWrite, MetadataVersion=3.4-IV0, Security=PLAINTEXT
    Build / JDK 11 and Scala 2.13 / org.apache.kafka.trogdor.coordinator.CoordinatorTest.testTaskRequestWithOldStartMsGetsUpdated()
```

**Pull request merged**
